taskt/Core/Automation/User32/GlobalHook.cs

```diff
@@ -0,0 +1,642 @@
+                _winEventHook = SetWinEventHook(SystemEvents.EventMin, SystemEvents.EventMax, IntPtr.Zero, _winEventHookHandler, 0, 0, 0);
+            }
+
+            //start stopwatch for timing all event occurences
+            _stopWatch = new Stopwatch();
+            _stopWatch.Start();
+
+            //stopwatch for tracking mouse moves specifically
+            _lastMouseMove = new Stopwatch();
+            _lastMouseMove.Start();
+
+        }
+
+        //hook end
+        public static void StopHook()
+        {
+            UnhookWindowsHookEx(_keyboardHookID);
+            UnhookWindowsHookEx(_mouseHookID);
+
+            if (_performWindowCapture)
+            {
+                UnhookWinEvent(_winEventHook);
+            }
+
+            //BuildCommentCommand();
+            HookStopped(null, new EventArgs());
+        }
+
+        //mouse and keyboard hook event triggers
+        private static IntPtr KeyboardHookEvent(int nCode, IntPtr wParam, IntPtr lParam)
+        {
+            if (nCode >= 0 && wParam == (IntPtr)WmKeyDown)
+            {
+                int vkCode = Marshal.ReadInt32(lParam);
+                BuildKeyboardCommand((Keys)vkCode);
+            }
+
+            return CallNextHookEx(_keyboardHookID, nCode, wParam, lParam);
+
+        }
+
+        public static event EventHandler<MouseCoordinateEventArgs> MouseEvent;
+
+        private static IntPtr MouseHookForLeftClickUpEvent(int nCode, IntPtr wParam, IntPtr lParam)
+        {
+            if (nCode >= 0)
+            {
+                var message = (MouseMessages)wParam;
+
+                if (message == MouseMessages.WmLButtonDown)
+                {
+                    if (StopOnClick)
+                    {
+                        UnhookWindowsHookEx(_mouseHookID);
+                    }
+
+                    //MSLLHOOKSTRUCT hookStruct = (MSLLHOOKSTRUCT)Marshal.PtrToStructure(lParam, typeof(MSLLHOOKSTRUCT));
+                    //System.Windows.Point point = new System.Windows.Point(hookStruct.pt.x, hookStruct.pt.y);
+
+                    System.Windows.Point point = new System.Windows.Point(Cursor.Position.X, Cursor.Position.Y);
+                    MouseEvent?.Invoke(null, new MouseCoordinateEventArgs() { MouseCoordinates = point });
+                }
+            }
+
+            return CallNextHookEx(_mouseHookID, nCode, wParam, lParam);
+        }
+
+        private static IntPtr MouseHookEvent(int nCode, IntPtr wParam, IntPtr lParam)
+        {
+            if (nCode >= 0)
+            {
+                BuildMouseCommand(lParam, (MouseMessages)wParam);
+            }
+
+            return CallNextHookEx(_mouseHookID, nCode, wParam, lParam);
+        }
+
+        //build keyboard command
+        private static void BuildKeyboardCommand(Keys key)
+        {
+            bool toUpperCase = false;
+
+            //determine if casing is needed
+            if (IsKeyDown(Keys.ShiftKey) && IsKeyToggled(Keys.Capital))
```

Comment:
Two final cases with reversed criteria can be less verbose: Later Pass

Reply:
Right